File: client/types/document.ts
export type DocumentCategory =
  | "lab_result"
  | "xray"
  | "prescription"
  | "insurance"
  | "vaccination_certificate"
  | "invoice"
  | "other";

export interface MedicalDocument {
  id: string;
  petId: string;
  ownerId: string;
  title: string;
  category: DocumentCategory;
  fileUrl: string;
  fileName: string;
  fileType: string;
  fileSizeBytes: number;
  notes: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface DocumentUploadPayload {
  petId: string;
  title: string;
  category: DocumentCategory;
  notes?: string;
  file: File;
}

export interface DocumentUpdatePayload {
  title?: string;
  category?: DocumentCategory;
  notes?: string;
}

interface RawMedicalDocument {
  id: string;
  pet_id: string;
  owner_id: string;
  title: string;
  category: DocumentCategory;
  file_url: string;
  file_name: string;
  file_type: string;
  file_size_bytes: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export function mapMedicalDocument(raw: RawMedicalDocument): MedicalDocument {
  return {
    id: raw.id,
    petId: raw.pet_id,
    ownerId: raw.owner_id,
    title: raw.title,
    category: raw.category,
    fileUrl: raw.file_url,
    fileName: raw.file_name,
    fileType: raw.file_type,
    fileSizeBytes: raw.file_size_bytes,
    notes: raw.notes,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

File: server/models/document.py
import uuid
from datetime import datetime, timezone
import enum

from sqlalchemy import String, DateTime, ForeignKey, Enum as SAEnum, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship

from server.database.base import Base


def _uuid() -> str:
    return str(uuid.uuid4())


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class DocumentCategory(str, enum.Enum):
    LAB_RESULT = "lab_result"
    XRAY = "xray"
    PRESCRIPTION = "prescription"
    INSURANCE = "insurance"
    VACCINATION_CERTIFICATE = "vaccination_certificate"
    INVOICE = "invoice"
    OTHER = "other"


class MedicalDocument(Base):
    __tablename__ = "medical_documents"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    pet_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("pets.id", ondelete="CASCADE"), nullable=False, index=True
    )
    owner_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    title: Mapped[str] = mapped_column(String(150), nullable=False)
    category: Mapped[DocumentCategory] = mapped_column(
        SAEnum(DocumentCategory, name="document_category"),
        default=DocumentCategory.OTHER,
        nullable=False,
    )
    file_url: Mapped[str] = mapped_column(String(500), nullable=False)
    file_name: Mapped[str] = mapped_column(String(255), nullable=False)
    file_type: Mapped[str] = mapped_column(String(50), nullable=False)
    file_size_bytes: Mapped[int] = mapped_column(Integer, nullable=False)
    notes: Mapped[str | None] = mapped_column(String(1000), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, onupdate=_utcnow, nullable=False
    )

    pet = relationship("Pet", backref="medical_documents")
